src/ext/ddtrace.c

```diff
@@ -175,6 +175,11 @@ static PHP_RSHUTDOWN_FUNCTION(ddtrace) {
         return SUCCESS;
     }
 
+    // todo: default DD_AUTOFINISH_SPANS to on for both userland and internally made spans
+    if (ddtrace_get_bool_config("DD_AUTOFINISH_SPANS", 0 TSRMLS_CC)) {
+        ddtrace_close_open_spans(TSRMLS_C);
```

Comment:
So far so great! :)

We might need to expose this function and env var check to userland since RSHUTDOWN is too late to close the spans. The tracer flushes in a `register_shutdown_function()` closure which runs [first in the shutdown process](https://github.com/php/php-src/blob/db233501ff9d56765ef4a870b777a643c2136711/main/main.c#L1832-L1835) so when [RSHUTDOWN is called](https://github.com/php/php-src/blob/db233501ff9d56765ef4a870b777a643c2136711/main/main.c#L1864-L1867), the tracer has already flushed.

Reply:
Yeah, I discovered this a bit later. I wonder if we ought to do it the other way, though: move our `register_shutdown_function` stuff into `RSHUTDOWN`? This would be more resilient in the face of someone using `exit` in a shutdown function, for instance. What do you think?

Reply:
I'm totally for this as we start moving things to the C level for sure. Right now about 90% of the flushing logic is in userland so we might need to keep it there for a while. Also - `register_shutdown_function()` closures are still called on `exit` so we wouldn't be necessarily gaining an edge there. But overall, I'm totally on board for eventually adding flush-on-RSHUTDOWN logic. :)

Reply:
Ah, I wasn't clear. In the event a shutdown handler calls exit then the remainder won't get called. It's a small edge, definitely. I was thinking about literally calling the PHP code we call in the shudown handler today from RSHUTDOWN. Basically, instead of an anonymous function we give it a name and then just call that from RSHUTDOWN. Thoughts?

Reply:
Ah! You are the edge-case master. Your original thought was clear - I just wasn't clear reading it. :)

I like the idea of flushing in userland on RSHUTDOWN. I've never tried running userland in RSHUTDOWN before and it's after the output buffer dies so I'm wondering if we might have to deal with some edge cases there. But I really love the idea! :) Are you thinking of exposing a userland function similar to `register_shutdown_function()` to register a closure that is called in RSHUTDOWN or something?

Reply:
Today we have this in `DDTrace\Bootstrap::tracerOnce`:

```php
    register_shutdown_function(function () {
        dd_trace_disable_in_request(); //disable function tracing to speedup shutdown

        $tracer = GlobalTracer::get();
        $scopeManager = $tracer->getScopeManager();
        $scopeManager->close();
        $tracer->flush();
    });
```

What if we took that anonymous function and put it in a static method, like `DDTrace\Bootstrap::shutdown_handler` or something, and then in RSHUTDOWN we did the internals equivalents of `class_exists('DDTrace\Bootstrap', $autoload = false)` and `method_exists('DDTrace\Bootstrap', 'shutdown_handler')` and if it is there then we call it?

Reply:
Even better! :)

Reply:
Also - the output buffer dies after RSHUTDOWN, not before as I stated earlier. So that's good! (I was looking at the post-RSHUTDOWN hook.)

Reply:
Current PR status: `Bootstrap::shutdown` is called from RSHUTDOWN on PHP 7 and on PHP 5 there is still `register_shutdown_handler` until I get a few more minutes to do the internals for it.